v2/tools/asoctl/internal/importing/resource_importer.go

```diff
@@ -36,91 +38,157 @@ func NewResourceImporter(
 	return &ResourceImporter{
 		scheme:    scheme,
 		client:    client,
+		pending:   make(map[string]ImportableResource),
 		completed: make(map[string]ImportableResource),
 	}
 }
 
 // Add adds an importer to the list of resources to import.
 func (ri *ResourceImporter) Add(importer ImportableResource) {
-	ri.pending = append(ri.pending, importer)
+	// Lock while we're modifying the maps
+	ri.lock.Lock()
+	defer ri.lock.Unlock()
+
+	ri.addCore(importer)
 }
 
 // AddARMID adds an ARM ID to the list of resources to import.
-func (ri *ResourceImporter) AddARMID(armID string) {
-	importer := NewImportableARMResource(armID, nil /* no owner */, ri.client, ri.scheme)
+func (ri *ResourceImporter) AddARMID(armID string) error {
+	importer, err := NewImportableARMResource(armID, nil /* no owner */, ri.client, ri.scheme)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create importer for %q", armID)
+	}
+
 	ri.Add(importer)
+	return nil
+}
+
+// addCore is the internal implementation of Add that assumes the lock is already held
+// This allows us to call it from other methods that already have the lock
+func (ri *ResourceImporter) addCore(importer ImportableResource) {
+	// If we've already handled this resource, skip it
+	if _, ok := ri.completed[importer.Name()]; ok {
+		return
+	}
+
+	// If we're already pending import of this resource, skip it
+	if _, ok := ri.pending[importer.Name()]; ok {
+		return
+	}
+
+	// Add it to our map and our queue
+	ri.queue = append(ri.queue, importer.Name())
+	ri.pending[importer.Name()] = importer
 }
 
 // Import imports all the resources that have been added to the importer.
 // Partial results are returned even in the case of an error.
 func (ri *ResourceImporter) Import(ctx context.Context) (*ResourceImportResult, error) {
-	var errs []error
-	var previousResource string
-	for len(ri.pending) > 0 {
-		// Remove the first pending importer
-		importer := ri.pending[0]
-		ri.pending = ri.pending[1:]
-
-		// If we've already handled this resource, skip it
-		if _, ok := ri.completed[importer.Name()]; ok {
-			continue
-		}
-
-		thisResource := len(ri.completed) + 1
-		pendingResources := len(ri.pending)
-		klog.Infof(
-			"Importing %d/%d: %s",
-			thisResource,
-			thisResource+pendingResources,
-			ri.idToLog(importer.Name(), previousResource))
+	processed := 0
+
+	// Why the nested loop?
+	// We can potentially trigger all the pending resources listed in ri.queue to be imported in parallel
+	// and end up with nothing on the queue because they're still running.
+	// The outer loop gives us a last look at the queue and allows us to keep going if there are new items
+	// requiring import.
+	for len(ri.queue) > 0 {
+		var eg errgroup.Group
+		eg.SetLimit(8)
```

Comment:
Why 8?

Reply:
Semi random number chosen so we get a reasonable degree of parallelism without trying to do too much. I'll add a comment.